orangecontrib/geo/widgets/owmap.py

```diff
@@ -1,906 +1,849 @@
-            return x, y
-
-    N_POINTS_PER_ITER = 1500
-
-    def redraw_markers_overlay_image(self, *args, new_image=False):
-        if not args and not self._drawing_args or self.data is None:
+        if t.n_loadings == 3:
+            self._show_internet_error(True)
             return
 
-        if args:
-            self._drawing_args = args
-        north, east, south, west, width, height, zoom, origin, map_pane_pos = self._drawing_args
-
-        lat, lon = self._latlon_data.T
-        visible = ((lat <= north) & (lat >= south) &
-                   (lon <= east) & (lon >= west)).nonzero()[0]
-        in_subset = (np.in1d(self.data.ids, self._subset_ids)
-                     if self._subset_ids.size else
-                     np.tile(True, len(lon)))
-
-        is_js_path = self.is_js_path = len(visible) < self.N_POINTS_PER_ITER
-
-        self._update_legend(is_js_path)
+        future = self.loader.get(t)
+        @future.add_done_callback
+        def set_tile(_future):
+            if _future.cancelled():
+                return
 
-        np.random.shuffle(visible)
-        # Sort points in subset to be painted last
-        visible = visible[np.lexsort((in_subset[visible],))]
+            assert _future.done()
 
-        if is_js_path:
-            self.evalJS('clear_markers_overlay_image()')
-            self._update_js_markers(visible, in_subset[visible])
-            self._owwidget.disable_some_controls(False)
-            return
+            if _future.exception():
+                _t = _future.exception().tile
+                _t.n_loadings += 1
+                # retry to download image
+                self._load_one_from_net(_t)
+            else:
+                _t, img = _future.result()
+                if not _t.disc_cache:
+                    self._show_internet_error(False)
+                self.mem_cache[_t] = img
+                self._add_patch(_t)
+                self._update_map_item()
+
+        self.futures.append(future)
+
+    def _show_internet_error(self, show):
+        if not self.master.Warning.no_internet.is_shown() and show:
+            self.master.Warning.no_internet()
+        elif self.master.Warning.no_internet.is_shown() and not show:
+            self.master.Warning.no_internet.clear()
+
+    def _cancel_futures(self):
+        for future in self.futures:
+            future.cancel()
+            if future._reply is not None:
+                future._reply.close()
+                future._reply.deleteLater()
+                future._reply = None
+
+    def set_tile_provider(self, tp):
+        self.tile_provider = tp
+        self.view_box.set_tile_provider(self.tile_provider)
+        self.tile_attribution.setHtml(self.tile_provider.attribution)
+
+    def clear_map(self):
+        self._cancel_futures()
+        self.futures = []
+        self.map = None
+        self.tz = 1
+
+        self.plot_widget.removeItem(self.b_map_item)
```

Comment:
This code is duplicated. Implement a function (i.e.) `update_map()`. And call it from `__init__` and `clear_map`.